### PR TITLE
(#110) magnetic cilia carpet

### LINKED ISSUE #110: Immersed magnetic cilia carpet illustration

### PR DESCRIPTION
**Pull request by sy-cui**
Fix #110. 

Magnetized cilia carpet under a rotating magnetic field. The velocity is averaged over each magnetization period, with time-averaged streamlines shown below.

https://user-images.githubusercontent.com/47127977/227403436-737ad978-ab0d-4bc9-9d7d-d37a99434b0f.mp4

### CODE REVIEW

File: examples/3d_examples/MagneticCiliaCarpetCase/immersed_magnetic_cilia_carpet.py [hunk 1]
```diff
@@ -171,9 +190,26 @@ def update_carpet_lag_grid_fields():
...
+            avg_vorticity *= 0.0
+            avg_velocity *= 0.0
+            no_period += 1
+
         # compute timestep
         flow_dt = flow_sim.compute_stable_timestep(dt_prefac=0.25)
 
+        # Average vorticity field
+        avg_vorticity += flow_sim.vorticity_field * flow_dt / period_timer_limit
+        avg_velocity += flow_sim.velocity_field * flow_dt / period_timer_limit
```
Comment (bhosale2) [inline examples/3d_examples/MagneticCiliaCarpetCase/immersed_magnetic_cilia_carpet.py:211]: Can you use an elementwise operation for this from pystencils? check out the saxpby kernel

File: examples/3d_examples/MagneticCiliaCarpetCase/immersed_magnetic_cilia_carpet.py [hunk 2]
```diff
@@ -219,18 +271,40 @@ def update_carpet_lag_grid_fields():
     carpet_base_centroid = np.array(
         [0.5 * domain_x_range, 0.5 * domain_y_range, 0.1 * domain_z_range]
     )
+    n_elem_per_rod = int(grid_size_x * rod_elem_prefactor / num_rods_along_x)
     cilia_carpet_simulator = MagneticCiliaCarpetSimulator(
+        magnetic_elastic_ratio=magnetic_elastic_ratio,
```
Comment (bhosale2) [inline examples/3d_examples/MagneticCiliaCarpetCase/immersed_magnetic_cilia_carpet.py:276]: lets call this magnetic Bond number?

File: examples/3d_examples/MagneticCiliaCarpetCase/magnetic_cilia_carpet.py [hunk 3]
```diff
@@ -7,12 +7,18 @@
 class MagneticCiliaCarpetSimulator:
     def __init__(
         self,
+        magnetic_elastic_ratio: float,  # MBAL2_EI, (Wang 2019)
```
Comment (bhosale2) [inline examples/3d_examples/MagneticCiliaCarpetCase/magnetic_cilia_carpet.py:10]: Lets call this magnetic bond number and maybe put a comment saying what it is.

File: examples/3d_examples/MagneticCiliaCarpetCase/magnetic_cilia_carpet.py [hunk 4]
```diff
@@ -26,10 +32,40 @@ class MagneticBeamSimulator(
...
+        moment_of_inertia = np.pi / 4 * base_radius**4
+        first_mode_frequency = (
+            3.5160
+            / rod_base_length**2
+            * np.sqrt(  # rad / s
+                youngs_modulus * moment_of_inertia / density / base_area
+            )
+        )
+
+        self.angular_frequency = frequency_ratio * first_mode_frequency
```
Comment (bhosale2) [inline examples/3d_examples/MagneticCiliaCarpetCase/magnetic_cilia_carpet.py:59]: a comment here for now, about this ratio, is would be helpful. we will later incorporate it into the documentation of the case.

**Review by bhosale2 — CHANGES_REQUESTED** [review]
Minor stuff, also please out a video or animation of the flow here.

File: examples/3d_examples/MagneticCiliaCarpetCase/immersed_magnetic_cilia_carpet.py (hunk 1 above)
Comment (sy-cui) [reply examples/3d_examples/MagneticCiliaCarpetCase/immersed_magnetic_cilia_carpet.py]: [fbd5513](https://github.com/SophT-Team/SophT/pull/120/commits/fbd5513ae66ab43926856f754aa07bef09e3ed80)

File: examples/3d_examples/MagneticCiliaCarpetCase/immersed_magnetic_cilia_carpet.py (hunk 2 above)
Comment (sy-cui) [reply examples/3d_examples/MagneticCiliaCarpetCase/immersed_magnetic_cilia_carpet.py]: [ce91966](https://github.com/SophT-Team/SophT/pull/120/commits/ce919664854a349d4b924b8ee527419355988c09)

File: examples/3d_examples/MagneticCiliaCarpetCase/magnetic_cilia_carpet.py (hunk 3 above)
Comment (sy-cui) [reply examples/3d_examples/MagneticCiliaCarpetCase/magnetic_cilia_carpet.py]: [ce91966](https://github.com/SophT-Team/SophT/pull/120/commits/ce919664854a349d4b924b8ee527419355988c09)

File: examples/3d_examples/MagneticCiliaCarpetCase/magnetic_cilia_carpet.py (hunk 4 above)
Comment (sy-cui) [reply examples/3d_examples/MagneticCiliaCarpetCase/magnetic_cilia_carpet.py:108]: Fixed in [9d5f50f](https://github.com/SophT-Team/SophT/pull/120/commits/9d5f50f36230aa8060c2efc016e620444b432a39)

**Pull request merged**